### ISSUE TITLE
[CSL-197] Improve peformance of crypto serialization & hashing 

### ISSUE THREAD

**Issue by jagajaga** [input-output-hk/cardano-sl#574]
<blockquote>@georgeee</blockquote>

Look into prof log posted in `cardano-sl-bnd` channel, `node1-2016-11-27.prof`. 4.8% CPU spent on hashing under `verifyBlock`, 24.4% spent on `ecPointFromOct` which is part of message deserialization. Later is due to fact we need to deserialize commitment even if we later decide we don't need it.

Proposal of this issue is to:
* add crypto types aliases kept as raw bytestrings/bytearrays
* add convenient convertors to existing types (with deserialization under the hood)
* use alias types whenever it's convenient (mostly communication)
* convert aliases to existing types on demand

This should be done only for types, used in communication (public keys, PVSS stuff). There is no need to do this for secret keys.

After completing issue, profile `cardano-sl` node in first minute of wroking to see load actually decreased.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

This issue is critical to finally get benchs working

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

It concerns only PVSS stuff, as it was recently discovered (Ed25519 library requires simple constructor to be applied)

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

@arodrigues to fix tests, after that task is to be closed

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

1. Why is this issue still `In Progress`? Tests were fixed long ago.
2. Did you do related benchmarks which give evidence that for all PVSS types encoding/decoding is expensive and for all other types it's not expensive? I guess no, but I think it makes sense to do them, because it may turn out that either for some PVSS types it doesn't make sense to store them as raw bytes (either because serialization is cheap or because we need to actually use them much more often than send or for some other reason) or that there are other types for which such change makes sense, but it's less evident. Anyway, I can't but agree that it's a separate task which is relatively minor.
3. I think we need tests which ensure that length of types for which we have lightweight wrappers is exactly what we expect (since these number are hardcoded in `SerTypes.hs`).
4. See my comment about in `SerTypes.hs`.
5. Wouldn't it be better to use a single type with phantom type parameter for all lightweight types instead of defining `newtype` for each type?

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

1. It's `Done` then
2. Yes, I tested. We could finally step to stable 100 with this fixed
Regarding other data types, `Ed25519` uses simple wrappers over ByteString. `Hash` is just `ByteArray`, which has slight overhead on deserialization cost (memory allocation) but not bigger. As to my analysis, it was only PVSS types that did some black magic costed us a fortune of CPU (when done for each SSC-related message, even one we would discard)
3. Check CSL-210
4. On GH? I replied already then
5. Well, it's possible. Didn't think on that.
And in most cases I want to check length of types. It would be cumbersome with phantom types

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

2. Ok, you convinced me to forget about it until we have related problems again. I hope it won't happen.
4. Yes, on GH. I wrote it just in case :) And anyway, I wrote one more comment which wasn't answered. If you don't mind, I will remove redundant (I am still sure it's absolutely redundant) `decodeFull`.
5. Ok, I will create a minor issue then.
> It would be cumbersome with phantom types
It will be less cumbersome than current solution.
